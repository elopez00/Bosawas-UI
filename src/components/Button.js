import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import colors from '../assets/colors.json'

/**
 * Primary button component
 * @param {Object} props - props pertaining to component 
 */
export default function Button(props) {
    // props 
    const { variant, disabled, onClick, children, href } = props;

    /**
     * Switches the type of button that will be displayed based off its variant prop
     * @returns { Component } - the styled component that corresponds to the variant
     */
    const switchVariants = () => {
        switch (variant) {
            case ('bubble') : return (
                <BubbleButton {...props} onClick={!disabled ? onClick : null}>
                    { children }
                </BubbleButton>
            );
            case ('outline') : return (
                <OutlineButton {...props} onClick={!disabled ? onClick : null}>
                    { children }
                </OutlineButton>
            )
            default : return (
                <DefaultButton {...props} onClick={!disabled ? onClick : null}>
                    { children }
                </DefaultButton>
            );
        }
    }

    return (
        <React.Fragment>
            { href ? <a href={!props.disabled ? href : null}>{switchVariants()}</a> : switchVariants() }
        </React.Fragment>
    )
}

Button.propTypes = {
    /**
     * needs to have some kind of content
     */
    children: PropTypes.any.isRequired,
    /**
     * overall style of the button  
     */ 
    variant: PropTypes.string,
    /**
     * overall color of the button  
     */ 
    color: PropTypes.string,
    /**
     * size of the button
     */
    size: PropTypes.string,
    /**
     * link that the button will redirect to
     */
    href: PropTypes.string,
    /**
     * handles button click
     */
    onClick: PropTypes.func,
    /**
     * disables button input
     */
    disabled: PropTypes.bool
}

Button.defaultProps = {
    color: "default",
    variant: "default",
    size: "default"
}

/**
 * Adjusts size based on the size prop
 * @param {Object} props - props pertaining to the styled component 
 */
const adjustPadding = props => {
    switch (props.size) {
        case ('large') : return '10px 20px';
        case ('small') : return '5px 10px';
        default : return '7.5px 15px';
    }
}

/**
 * Adjusts text color based on the color prop
 * @param {Object} props - props pertaining to the styled component 
 */
const adjustTextColor = props => {
    if (colors[props.color]) {
        return props.color === 'default' ? colors.default.color : colors[props.color].dark;
    } else {
        return colors.default.color;
    }
}

// styled components
const DefaultButton = styled.button`
    background: ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
    color: ${props => colors[props.color] ? colors[props.color].color : colors.default.color};
    font-size: 1rem;
    padding: ${props => adjustPadding(props)};
    border: none;
    border-radius: 5px;
    outline: none;
    cursor: pointer;
    transition: background 300ms, box-shadow 150ms;
    box-shadow: 0 0 0 0 transparent;
    &:hover {
        background: ${props => colors[props.color] ? colors[props.color].dark : colors.default.dark};
    }
    &:active {
        box-shadow: 0 0 0 3px ${props => colors[props.color] ? colors[props.color].glow : colors.default.glow};
    }
    &:disabled {
        color: #6b6b6b;
        background: #d9d9d9;
        cursor: default;
    }
`
const OutlineButton = styled.button`
    background: transparent;
    color: ${props => adjustTextColor(props)};
    font-size: 1rem;
    border-radius: 5px;
    padding: ${props => adjustPadding(props)};
    border: 1px solid ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
    outline: none;
    transition: background 300ms, color 300ms, border 300ms;
    cursor: pointer;
    &:hover {
        border: 1px solid ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.default.glow};
        color: ${props => adjustTextColor(props)};
    }
    &:active {
        border: 1px solid ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
        background: ${props => colors[props.color] ? colors[props.color].intenseGlow : colors.default.intenseGlow};
        color: ${props => {
            if (colors[props.color]) {
                return props.color === 'default' ? colors.default.color : colors[props.color].dark;
            } else {
                return colors.default.color;
            }
        }};
    }
    &:disabled {
        color: ${colors.default.dark};
        border: 1px solid ${colors.default.default};
        cursor: default;
        &:hover {
            border: 1px solid ${colors.default.default};
            background: transparent;
        }
    }
`
const BubbleButton = styled.button`
    background: transparent;
    color: inherit;
    font-size: 1rem;
    border: none;
    border-radius: 100px;
    padding: ${props => adjustPadding(props)};
    outline: none;
    border-radius: 100px;
    cursor: pointer;
    transition: background 300ms, color 300ms;
    &:hover {
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.default.glow}; 
        color: ${props => adjustTextColor(props)};
    }
    &:active {
        background: ${props => colors[props.color] ? colors[props.color].intenseGlow : colors.default.intenseGlow};
        color: ${props => adjustTextColor(props)};
    }
    &:disabled {
        color: #6b6b6b;
        cursor: default;
        &:hover {
            background: transparent;
        }
    }
`