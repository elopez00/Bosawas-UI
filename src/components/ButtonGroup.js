import React from 'react'
import styled, {css} from 'styled-components'
import colors from '../assets/colors.json'
import PropTypes from 'prop-types'
import findByType from './helper/findByType'

// initializes replica button
const Button = () => null;
Button.displayName = 'Button'

/**
 * Displays button group component
 * @param {Object} props - props pertainign to component 
 */
export default function ButtonGroup(props) {
    const { children, color, size, variant } = props;

    /**
     * Switches the type of button that will be displayed based off its variant prop
     * @returns { Component } - the styled component that corresponds to the variant
     */
    const renderButtons = () => {
        const buttons = findByType(children, Button);
        if (!buttons.length) return;

        let index = -1;
        return buttons.map(button => { 
            index ++; 
            switch (variant) {
                case ('outline') : return (
                    <OutlineButton
                     {...button.props}
                     key={Math.random()}
                     color={color}
                     size={size}
                     edgeType={button.length !== 1 ? !index ? 'left' : index === buttons.length-1 ? 'right' : 'middle' : 'all'}>
                         { button.props.children }
                    </OutlineButton>
                );
                case ('bubble') : return (
                    <BubbleButton
                     {...button.props}
                     key={Math.random()}
                     color={color}
                     size={size}
                     edgeType={button.length !== 1 ? !index ? 'left' : index === button.length-1 ? 'right' : 'middle' : 'all'}>
                        { button.props.children }
                    </BubbleButton>
                )
                default : return (
                    <DefaultButton 
                     {...button.props} 
                     key={Math.random()} 
                     color={color} 
                     size={size}
                     edgeType={button.length !== 1 ? !index ? 'left' : index === buttons.length-1 ? 'right' : 'middle' : 'all'}>
                        {button.props.children }
                    </DefaultButton>
                );
            }
            
        })
    }

    return (
        <CButtonGroup {...props}>
            { renderButtons() }
        </CButtonGroup>
    )
}

ButtonGroup.propTypes = {
    // is the size of the buttons
    size: PropTypes.string,
    // is the style of the button
    variant: PropTypes.string,
    // is the color of the buttons
    color: PropTypes.string,
}

ButtonGroup.defaultProps = {
    color: "default",
    size: "default",
    variant: "default"
}

/**
 * Adjusts size based on the size prop
 * @param {Object} props - props pertaining to the styled component 
 * @returns {String} - padding settings
 */
const adjustPadding = props => {
    switch (props.size) {
        case ('large') : return '10px 20px';
        case ('small') : return '5px 10px';
        default : return '7.5px 15px';
    }
}

/**
 * Determines which is the order of the buttons
 * @param {Object} props - props pertaining to the styled component
 * @returns {String} - border settings;
 */
const adjustBorderRadius = props => {
    switch (props.edgeType) {
        case ('right') : return '0 5px 5px 0';
        case ('left') : return '5px 0 0 5px';
        case ('all') : return '5px';
        default : return '0';
    }
}

/**
 * Adjusts text color based on the color prop
 * @param {Object} props - props pertaining to the styled component 
 * @returns {String} - style for text
 */
const adjustTextColor = props => {
    if (colors[props.color]) {
        return props.color === 'default' ? colors.default.color : colors[props.color].dark;
    } else {
        return colors.default.color;
    }
}

/**
 * Adjusts the border for default buttons
 * @param {Object} props - props pertaining to the styled component
 * @returns {String} = style for border
 */
const editDefaultBorder = props => {
    let borderColor = colors[props.color] ? colors[props.color].dark : colors.default.dark;
    if (props.edgeType === 'left' || props.edgeType === 'middle') return `1px solid ${borderColor}`;
    else return 'none';
}

/**
 * Determines the the border layout of the outline buttons
 * @param {Object} props - props pertaining to the styled component
 * @returns {String} - special template string containg css that is easily integrated to existing 
 * css in the styled component initialization 
 */
const adjustOultineBorder = props => {
    let borderColor = colors[props.color] ? colors[props.color].default : colors.default.default;
    switch (props.edgeType) {
        case ('left') : return css`
            border: 1px solid;
            border-radius: 5px 0 0 5px;`
        case ('middle') : return css`
            border-top: 1px solid;
            border-bottom: 1px solid;
            border-right: 1px solid;
            border-left: none;
            border-radius: 0;`
        case ('right') : return css`
            border-top: 1px solid;
            border-bottom: 1px solid;
            border-right: 1px solid;
            border-left: none;
            border-radius: 0 5px 5px 0;`
        default : return css`
            border: 1px solid;
            border-radius: 5px;`
    }
}


// styled components
const CButtonGroup = styled.div`
    
`
const DefaultButton = styled.button`
    padding: ${props => adjustPadding(props)};
    font-size: 1rem;
    background: ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
    color: ${props => colors[props.color] ? colors[props.color].color : colors.default.color};
    border: none;
    border-right: ${props => editDefaultBorder(props)};
    border-radius: ${props => adjustBorderRadius(props)};
    transition: background 300ms;
    cursor: pointer;
    outline: none;
    &:hover {
        background: ${props => colors[props.color] ? colors[props.color].dark : colors.default.dark};
    }
`
const OutlineButton = styled.button`
    padding: ${props => adjustPadding(props)};
    font-size: 1rem;
    background: transparent;
    color: ${props => adjustTextColor(props)};
    ${props => adjustOultineBorder(props)};
    border-color: ${props => colors[props.color] ? colors[props.color].default : colors.default.default};
    cursor: pointer;
    transition: background 300ms;
    outline: none;
    &:hover {
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.default.glow}
    }
`
const BubbleButton = styled.button`
    padding: ${props => adjustPadding(props)};
    font-size: 1rem;
    color: inherit;
    border-radius: 100px;
    border: none;
    outline: none;
    background: transparent;
    cursor: pointer;
    transition: background 300ms, color 300ms;
    &:hover {
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.default.glow};
        color: ${props => adjustTextColor(props)};
    }

`