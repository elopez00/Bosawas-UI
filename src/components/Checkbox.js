import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import colors from '../assets/colors.json'

/**
 * This component displays the checkbox
 * @param {Object} props - props pertaining to component
 */
export default function Checkbox(props) {
    // props 
    const { color, labelStyle, label, disabled } = props;
    return (
        <Label color={color} style={labelStyle} disabled={disabled}>
            <CCheckbox type="checkbox" {...props}/>
            {label}
            <i className="material-icons">check</i>
        </Label>
    )
}

Checkbox.propsTypes = {
    /**
     * Disabled checkbox
     */
    disabled: PropTypes.bool
}

// styled components
const Label = styled.label`
    position: relative;
    width: auto;
    color: ${props => props.disabled ? "#6b6b6b" : "inherit" };
    transition: color 300ms;
    & > i {
        font-size: 1rem;
        position: absolute;
        top: 1px;
        left: 5px;
        display: none;
        cursor: pointer;
        color: ${props => colors[props.color] ? colors[props.color].color : colors.blue.color};
    }
    & input:checked ~ i {
        display: block;
    }
`
const CCheckbox = styled.input`
    -webkit-appearance: none;
	background: ${colors.default.default};
	padding: 9px;
	border-radius: 3px;
	display: inline-block;
	position: relative;
    cursor: pointer;
    margin-right: 5px;
    transform: translateY(-2px);
    vertical-align: middle;
    outline: none;
    transition: background 100ms, border 300ms, color 100ms;
    &:hover {
        transition: background 200ms;
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.blue.glow}
    }
    &:active {
        background: ${props => colors[props.color] ? colors[props.color].intenseGlow : colors.blue.intenseGlow}
    }
    &:checked:active {
        background: ${props => colors[props.color] ? colors[props.color].intenseGlow : colors.blue.intenseGlow};
    }
    &:checked {
        background: ${props => colors[props.color] ? colors[props.color].default : colors.blue.default};
        color: ${props => colors[props.color] ? colors[props.color].color : colors.blue.color};
    }
    &:disabled {
        color: #6b6b6b;
        background: #d9d9d9;
        cursor: default;
    }
`