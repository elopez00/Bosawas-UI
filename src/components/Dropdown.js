import React, { Component } from 'react'
import { Button as CButton, Option as COption } from './'
import styled from 'styled-components'
import findByType from './helper/findByType'
import PropTypes from 'prop-types'
import colors from '../assets/colors.json'

// initializes replica Option
const Option = () => null;
Option.displayName = "Option";

// initializes replica Button
const Button = () => null;
Button.displayName = "Button";

class Dropdown extends Component {
    // state
    state = {
        menuHeight: 0,
        menuOpen: false,
        overflow: false
    }

    // references
    dropdown = React.createRef();
    menu = React.createRef();

    /**
     * adds event listener that checks if mouse press occurrs outside menu
     */
    componentDidMount() {
        document.addEventListener('mousedown', this.handleOutsideClick);
        this.setState({menuHeight: this.menu?.current?.scrollHeight});
    }   

    /**
     * removes the event listener before component unmounts
     */
    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleOutsideClick)
    }

    /**
     * handles click outside of the dropdown menu
     * @param {Object} event - mousedown event 
     */
    handleOutsideClick = event => {
        const { overflow } = this.state;
        if (this.dropdown?.current && !this.dropdown?.current?.contains(event.target)) {
            this.setState({menuOpen: false});
            overflow && this.setState({overflow: false})
        }
    }

    /**
     * handles button click
     * @param {Object} props - props pertaining to the Button subcomponent
     */
    handleClick = props => {
        const { menuOpen, overflow } = this.state;
        props.onClick && props.onClick();
        this.setState({menuOpen: !menuOpen});
        if (!overflow) {
            setTimeout(() => this.setState({overflow: !overflow}), 300)
        } else {
            this.setState({overflow: !overflow});
        }
    }

    /**
     * Handles the onClick functionality for the Option subcomponent
     * @param {Object} props - props pertaining to Option subcomponent 
     */
    handleOptionClick = props => {
        const { onClick } = props;
        const { menuOpen } = this.state;
        onClick && onClick();
        this.setState({
            overflow: false,
            menuOpen: !menuOpen,
            menuHeight: this.menu.current.scrollHeight
        })
    }

    /**
     * Renders Dropdown button
     * @returns {Component} - DropdownButton
     */
    renderButton = () => {
        const { children } = this.props;
        const { menuOpen } = this.state;
        const button = findByType(children, Button)[0];
        if (!button) return;

        return <CButton {...button.props} onClick={() => this.handleClick(button.props)}>
            {button.props.children}
            <Carat open={menuOpen}>
                <i className="material-icons">keyboard_arrow_down</i>
            </Carat>
        </CButton>
    }

    /**
     * Options for the select menu
     * @returns {Component} - Option styled component rendered in select menu
     */
    renderOptions = () => {
        const { children } = this.props;
        const options = findByType(children, Option);
        if (!options.length) return;

        return options.map(option => (
            <COption 
             key={Math.random()} 
             {...option.props} 
             onClick={() => this.handleOptionClick(option.props)}>
                { option.props.children }
            </COption>
        ));
    }

    render() {
        const { menuHeight, menuOpen, overflow } = this.state;
        return (
            <CDropdown
             height={menuHeight} 
             open={menuOpen} 
             ref={this.dropdown}  
             {...this.props}>
                { this.renderButton() }
                <Options 
                 height={menuHeight} 
                 open={menuOpen} 
                 ref={this.menu} 
                 menuOverflow={overflow}>
                    { this.renderOptions() }
                </Options>
            </CDropdown>
        )
    }
}

export default Dropdown;

// styled components
const CDropdown = styled.div`
    display: inline-block;
    position: relative;
    & > button {
        display: flex;
        align-items: center;
        justify-content: space-between;
    }
`
const Carat = styled.a`
    display: inline-block;
    display: flex;
    align-items: center;
    transform: translate(6px);
    & i {
        transition: transform 300ms;
        transform: ${props => props.open ? 'rotate(180deg)' : 0};
    }
`
const Options = styled.div`
    display: inline-block;
    overflow: ${props => props.menuOverflow ? 'auto' : 'hidden'};
    position: absolute;
    top: 110%;
    left: 0;
    min-width: auto;
    width: 100%;
    height: ${props => props.open ? props.height : 0}px;
    transition: height 300ms;
    max-height: ${props => props.maxHeight || 500}px;
    background: white;
    border-radius: 5px;
    box-shadow: 0px 2px 5px rgba(0, 0, 0, .2);
    z-index: 1;
`