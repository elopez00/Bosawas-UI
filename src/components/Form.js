import React from 'react'
import styled from 'styled-components'
import findAll from './helper/findAll'
import {
    TextInput as CTextInput,
    Switch as CSwitch,
    Radio as CRadio,
    Checkbox as CCheckbox,
    Select as CSelect
} from './'
import PropTypes from 'prop-types';

export default function Form(props) {
    /**
     * renders all text inputs
     * @param {Object} textInputs - object containing text input components and indexes
     */
    const renderTextInputs = textInputs => textInputs?.map(textInput => {
        const { child, index } = textInput;
        return { 
            child: <CTextInput {...child.props} key={Math.random()}>{child.props.children}</CTextInput>,
            index 
        }
    })
    
    /**
     * renders all switches
     * @param {Object} switches - object containing switch components and indexes
     */
    const renderSwitches = switches => switches?.map(cswitch => {
        const { child, index } = cswitch;
        return { 
            child: <CSwitch {...child.props} key={Math.random()}/>,
            index
        }
    })

    /**
     * renders all radios
     * @param {Object} radios - object containing radio components and indexes
     */
    const renderRadios = (radios, group) => radios?.map(radio => {
        const { child, index } = radio;
        return { 
            child: <CRadio {...child.props} name={group || child.props.name} key={Math.random()} />,
            index
        }
    })

    /**
     * renders all checkboxes
     * @param {Object} checkboxes - object containing checkbox components and indexes 
     */
    const renderCheckboxes = (checkboxes, group) => checkboxes?.map(checkbox => {
        const { child, index } = checkbox;
        return {
            child: <CCheckbox {...child.props} name={group || child.props.name} key={Math.random()} />,
            index
        }
    })

    /**
     * renders all selects
     * @param {Object} selects - object containing all select components and indexes 
     */
    const renderSelects = selects => selects?.map(select => {
        const { child, index } = select;
        return {
            child: <CSelect {...child.props} key={Math.random()} />, 
            index
        }
    })

    /** 
     * renders all groups
     * @param {Objects} groups - object containing all select compnents and indexes 
     */
    const renderGroups = groups => groups?.map(group => {
        const { child, index } = group;
        return {
            child: <FormGroup {...child.props} key={Math.random()}>
                { child.props.title ? <h3>{child.props.title}</h3> : null}
                { renderContent(child.props) }
            </FormGroup>,
            index
        }
    })

    /**
     * Renders all children with speacial components
     * @param {Object} props - props pertaining to the component being rendered 
     */
    const renderContent = cprops => {
        // components
        const components = findAll(cprops.children, [TextInput, Switch, Radio, Checkbox, Select, Group]);

        // individual components
        const TextInputs = renderTextInputs(components['TextInput']) || [];
        const Switches = renderSwitches(components['Switch']) || [];
        const Radios = renderRadios(components['Radio'], cprops.name) || [];
        const Checkboxes = renderCheckboxes(components['Checkbox'], cprops.name) || [];
        const Selects = renderSelects(components['Select']) || [];
        const Groups = renderGroups(components['Form.Group']) || [];
        const Others = components['other'] || [];
        const renderAll = [...TextInputs, ...Switches, ...Radios, ...Checkboxes, ...Selects, ...Groups, ...Others]
            .sort((a, b) => a.index - b.index);

        return renderAll.map(component => component.child)
    }

    return (
        <CForm {...props}>
            { renderContent(props) }
        </CForm>
    )
    
}

// create prototype components
const TextInput = () => null;
const Switch = () => null;
const Radio = () => null;
const Checkbox = () => null;
const Select = () => null;
const Group = () => null;

// set respective display names
TextInput.displayName = 'TextInput';
Switch.displayName = 'Switch';
Radio.displayName = 'Radio';
Checkbox.displayName = 'Checkbox';
Select.displayName = 'Select'
Group.displayName = 'Form.Group'

// assign group as subcomponent
Form.Group = Group;

// prop types
Form.propTypes = {
    /** Title of form*/
    title: PropTypes.string,
    /** Width of form */
    width: PropTypes.string,
    /** onSubmit */
    onSubmit: PropTypes.func,
    
}

const CForm = styled.form`
    display: inline-flex;
    flex-direction: column;
    width: ${props => props.width || 'auto'};
    & > label {
        margin: 10px;
    }
    & input[type="text"] {
        flex: 1;
    }
`
const FormGroup = styled.div`
    flex: 1;
    display: flex;
    flex-direction: ${props => props.split ? 'auto' : 'column'};
    margin-bottom: 15px;
    & label {
        margin: 5px 10px;
        flex: 1;
    }
    & input[type="text"] {
        width: 100%;
    }
    & h3 {
        margin: 0;
        margin-bottom: 15px;
    }
`