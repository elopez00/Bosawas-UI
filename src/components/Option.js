import React from 'react'
import styled from 'styled-components'
import colors from '../assets/colors.json'
import PropTypes from 'prop-types'

/**
 * Option for dropdown menus 
 * @param {Object} props - props pertaining to object 
 */
export default function Option(props) {
    const { children } = props;
    return (
        <COption {...props}>{children}</COption>
    )
}

Option.propTypes = {
    // children must be string
    children: PropTypes.string.isRequired
}

// styled components
const COption = styled.span`
    display: block;
    padding: 8px;
    font-size: 1rem;
    transition: background 300ms;
    cursor: default;
    &:hover {
        background: ${colors.default.dark};
    }
`