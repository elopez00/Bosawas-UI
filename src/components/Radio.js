import React from 'react';
import styled from 'styled-components'
import colors from '../assets/colors.json'
import PropTypes from 'prop-types'

/**
 * Component for the Radio input
 * @param {Object} props - props pertaining to the component 
 */
export default function Radio(props) {
    // props 
    const { label, disabled } = props; 
    return (
        <Label disabled={disabled}>
            <CRadio type="radio" {...props} />
            <div>{label}</div>
        </Label>
    )
}
Radio.propTypes = {
    // the group which the radio belongs to
    name: PropTypes.string,
    // boolean determining a default check in the radio
    defaultChecked: PropTypes.bool,
    // disables component
    disabled: PropTypes.bool
}

Radio.defaultProps = {
    background: "default",
    size: "default"
}

// styled components
const Label = styled.label`
    display: inline-flex;
    & > div {
        transform: translateY(2px)
    }
    color: ${props => props.disabled ? "#6b6b6b" : "inherit"};
`

const CRadio = styled.input`
    -webkit-appearance: none;
    vertical-align: -4px;
    margin-right: 5px;
    background: ${colors.default.default};
    padding: 8px;
    border-radius: 50%;
    outline: none;
    position: relative;
    cursor: pointer;
    &:hover {
        transition: background 300ms;
        background: ${props => colors[props.color] ? colors[props.color].glow : colors.blue.glow }
    }
    &:active {
        background: ${props => colors[props.color] ? colors[props.color].intenseGlow : colors.blue.intenseGlow}
    }
    &:checked {
        background: ${props => colors[props.color] ? colors[props.color].default : colors.blue.default}
    }
    &:checked:after {
        content: '';
        position: absolute;
        padding: 4px;
        background: white;
        top: 4px;
        left: 4px;
        border-radius: 50%;
    }
    &:disabled {
        color: #6b6b6b;
        background: #d9d9d9;
        cursor: default;
    }
`