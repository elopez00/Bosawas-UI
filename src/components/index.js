export {default as Button} from './Button'
export {default as ButtonGroup} from './ButtonGroup'
export {default as Checkbox} from './Checkbox'
export {default as Radio} from './Radio'