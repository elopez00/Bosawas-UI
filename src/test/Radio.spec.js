import React from 'react'
import { Radio } from '../components'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

describe ('<Radio/>', () => {
    // snapshot
    it ('renders correctly', () => {
        const tree = renderer
            .create(<Radio label="This is a radio" defaultChecked/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

    // render testing
    it ('renders radio input element', () => {
        const wrapper = mount (<Radio />);
        expect(wrapper
            .find('input')
            .props()
            .type
        ).toBe('radio')
    })

    it ('renders label correctly', () => {
        const wrapper = mount (<Radio label="this is a label"/>);
        expect(wrapper
            .find('input')
            .props()
            .label
        ).toBe('this is a label');
        expect(wrapper.text()).toBe('this is a label');
    })

    // prop testing
    it ('successfully fires onChange function on check', () => {
        const mockChange = jest.fn();
        const wrapper = mount (<Radio onChange={mockChange}/>);
        wrapper.find('input').simulate('change', { target: { checked: true } });
        expect(mockChange).toHaveBeenCalled();
    })

    it ('does not call onClick function when disabled=true', () => {
        const mockClick = jest.fn();
        const wrapper = mount (<Radio onClick={mockClick} disabled/>);
        wrapper.find('input').simulate('click');
        expect(mockClick).toHaveBeenCalledTimes(0);
    })
})