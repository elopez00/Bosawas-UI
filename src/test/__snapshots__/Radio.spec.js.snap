// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio/> renders correctly 1`] = `
<label
  className="Radio__Label-pzn6oc-0 eQXWpR"
>
  <input
    className="Radio__CRadio-pzn6oc-1 cnuZUk"
    defaultChecked={true}
    label="This is a radio"
    size="default"
    type="radio"
  />
  <span>
    This is a radio
  </span>
</label>
`;
